### PR TITLE
[LG-5455] feat(code-editor): add download functionality with smart file extensions

### PR DESCRIPTION
**Pull request by tsck**
## ✍️ Proposed changes

This PR adds comprehensive download functionality to the CodeEditor component, allowing users to download their code with appropriate file extensions based on the selected language. The implementation includes both imperative API access and Panel UI integration.

**Key Features:**
- **Smart file extensions**: Automatic mapping from editor language to appropriate file extension (`.js`, `.ts`, `.py`, `.css`, etc.) with `.txt` fallback
- **Imperative API**: New `downloadContent(filename?)` method on CodeEditor ref handle
- **Panel integration**: Download button in secondary menu now performs actual file downloads
- **Content validation**: Prevents downloads of empty/whitespace-only content with user warnings
- **Context support**: Download function available to child components via CodeEditorContext
- **Comprehensive testing**: Behavior-focused tests that verify actual download triggers rather than implementation details

The implementation uses browser-native Blob API and maintains full backward compatibility while extending the component's capabilities. Supports 16+ languages including JavaScript, TypeScript, Python, Java, Go, Rust, and more.

🎟️ _Jira ticket:_ [LG-5455](https://jira.mongodb.org/browse/LG-5455)

### For bug fixes, new features & breaking changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have run `npm run build:ts` and there are no errors
- [x] I have run `npm test` and there are no errors
- [x] I have run `npm run lint:eslint` and there are no errors  
- [x] I have run `npm run lint:stylelint` and there are no errors
- [x] I have added or updated a Changeset (`npx changeset`)
- [ ] I have tested this change in Storybook and/or in a test application
- [ ] I have tested this change against both light and dark mode (if applicable)

### Testing Download Functionality
1. **Basic Download Test:**
   - Create a CodeEditor with some content: `<CodeEditor defaultValue="console.log('test');" language="javascript" />`
   - Call `ref.current.downloadContent('my-script')` or use Panel download button
   - Verify file downloads as `my-script.js` with correct content

2. **Language Extension Mapping:**
   - Test different languages: JavaScript (`.js`), TypeScript (`.ts`), Python (`.py`), etc.
   - Verify correct file extensions are applied automatically
   - Test with no language specified - should default to `.txt`

3. **Edge Cases:**
   - Try downloading empty content - should show warning and prevent download
   - Try downloading whitespace-only content - should show warning and prevent download
   - Test default filename behavior when no filename provided

4. **Panel Integration:**
   - Use `<Panel showSecondaryMenuButton />` component
   - Click secondary menu → Download option
   - Verify download is triggered through context

5. **Imperative API:**
   - Test `ref.current.downloadContent()` without filename (uses 'code' default)
   - Test `ref.current.downloadContent('custom-name')` with custom filename
   - Verify both methods work correctly

### Testing in Storybook
- Navigate to CodeEditor stories
- Test download functionality in different language configurations
- Verify Panel component download integration works properly

### CODE REVIEW

File: .changeset/honest-clocks-join.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+- **Smart file extensions**: Automatic file extension mapping based on editor language (`.js`, `.ts`, `.py`, `.css`, etc.) with fallback to `.txt`
+- **Panel integration**: Download button in Panel secondary menu now performs actual file downloads with proper filename handling
+- **Context support**: Download function available to child components via `CodeEditorContext` for custom UI implementations
+- **Content validation**: Prevents downloads of empty or whitespace-only content with user-friendly warnings
+- **Enhanced TypeScript**: Complete `downloadContent` method documentation in `CodeEditorHandle` interface
+- **Comprehensive testing**: Behavior-focused tests that verify actual download triggers rather than implementation details
+- **Language support**: Built-in extension mapping for 16+ languages including JavaScript, TypeScript, Python, Java, Go, Rust, and more
+
+**Technical implementation**: Uses browser-native Blob API and temporary anchor elements for downloads. Maintains full backward compatibility while adding new download capabilities. Language extension mapping ensures downloaded files have appropriate extensions for syntax highlighting in external editors.
```
Comment (shaneeza) [inline .changeset/honest-clocks-join.md:16]: TIL Blob API!

File: packages/code-editor/src/CodeEditor/CodeEditor.tsx [hunk 2]
```diff
@@ -35,6 +35,28 @@ import {
   useModuleLoaders,
 } from './hooks';
 
+/**
+ * Mapping of language names to their corresponding file extensions
+ */
+const LANGUAGE_EXTENSION_MAP = {
```
Comment (shaneeza) [inline packages/code-editor/src/CodeEditor/CodeEditor.tsx]: Is there a constants file this can be moved to?

File: packages/code-editor/src/CodeEditor/CodeEditor.tsx [hunk 3]
```diff
@@ -179,6 +201,51 @@ export const CodeEditor = forwardRef<CodeEditorHandle, CodeEditorProps>(
...
+      (filename = 'code'): void => {
+        const content = getContents();
+
+        if (!content.trim()) {
+          console.warn('Cannot download empty content');
+          return;
+        }
+
+        // Determine file extension based on language
+        const extension =
```
Comment (shaneeza) [inline packages/code-editor/src/CodeEditor/CodeEditor.tsx]: Can someone pass in a language that isn't included in `LANGUAGE_EXTENSION_MAP`?

File: packages/code-editor/src/Panel/Panel.tsx [hunk 4]
```diff
@@ -115,6 +116,13 @@ export function Panel({
     onRedoClick?.();
   };
 
+  const handleDownloadClick = () => {
+    if (downloadContent) {
+      downloadContent();
+    }
+    onDownloadClick?.();
```
Comment (shaneeza) [inline packages/code-editor/src/Panel/Panel.tsx:125]: Is there a test that tests that `onDownloadClick` is called?

File: packages/code-editor/README.md [hunk 5]
```diff
@@ -321,8 +328,25 @@ The `CodeEditorHandle` provides the following methods:
...
+The `downloadContent` method automatically determines the appropriate file extension based on the editor's language:
+
+```tsx
+// Downloads as 'script.js' (JavaScript extension)
+editorRef.current.downloadContent('script');
+
+// Downloads as 'code.py' (default filename with Python extension)
+editorRef.current.downloadContent(); // No filename provided
+
+// Downloads as 'document.txt' (fallback for unsupported/no language)
```
Comment (shaneeza) [inline packages/code-editor/README.md]: Is this missing a code snippet?

File: packages/code-editor/src/Panel/Panel.tsx [hunk 6]
```diff
@@ -115,6 +116,13 @@ export function Panel({
     onRedoClick?.();
   };
 
+  const handleDownloadClick = () => {
+    if (downloadContent) {
+      downloadContent();
```
Comment (shaneeza) [inline packages/code-editor/src/Panel/Panel.tsx]: If you're downloading from the panel, will the file name always be `code`?

File: packages/code-editor/src/CodeEditor/CodeEditor.tsx (hunk 2 above)
Comment (tsck) [reply packages/code-editor/src/CodeEditor/CodeEditor.tsx]: There is not, but I'll make one and put this in there!

File: packages/code-editor/src/CodeEditor/CodeEditor.tsx (hunk 3 above)
Comment (tsck) [reply packages/code-editor/src/CodeEditor/CodeEditor.tsx]: No, this is a great point. I'll also update the `LANGUAGE_EXTENSION_MAP` to enforce this.

File: packages/code-editor/src/Panel/Panel.tsx (hunk 6 above)
Comment (tsck) [reply packages/code-editor/src/Panel/Panel.tsx]: Currently yes, but I think this is a good point. I'm going to add an optional prop on `Panel` that allows the setting of it.

Comment (tsck) [reply packages/code-editor/src/Panel/Panel.tsx]: I also decided I'm going to update `handleDownloadContent` in CodeEditor, to only add an extension if an acceptable extension is already added to the filename provided. This feels intuitive to me. So if I have `js` selected and pass in a filename of `index.js` it doesn't download as `index.js.js` but if my filename is just `index` it will download as `index.js` intelligently.

I also think I'm going to get rid of the default `.txt` because its probably not a fair assumption. When no language is selected, just no extension will be added.

Curious to hear your thoughts on this.

Comment (tsck) [reply packages/code-editor/src/Panel/Panel.tsx]: I'm backtracking on this lol. I'm now thinking that a filename provided should override the extension. It just gives the consumer more control over the download. Still curious what you think here though

File: packages/code-editor/src/Panel/Panel.tsx (hunk 4 above)
Comment (tsck) [reply packages/code-editor/src/Panel/Panel.tsx:125]: Yup, it's called `calls onDownloadClick when download menu item is clicked` in `Panel.spec.tsx`

File: packages/code-editor/README.md (hunk 5 above)
Comment (tsck) [reply packages/code-editor/README.md]: Good catch. I removed this completely since I'm no longer defaulting to `.txt` at the moment

File: packages/code-editor/src/Panel/Panel.tsx (hunk 6 above)
Comment (shaneeza) [reply packages/code-editor/src/Panel/Panel.tsx]: if no extension is specified, what will the format be?

Comment (shaneeza) [reply packages/code-editor/src/Panel/Panel.tsx]: I was just testing out preview on mac, https://jmp.sh/EWQKXqlV. I'm not sure if these are relatable but if you don't add an extention, it adds one by default, and if you type in an extension, it treats it as part of the name.

File: packages/code-editor/README.md [hunk 7]
```diff
@@ -263,8 +272,30 @@ The `CodeEditorHandle` provides the following methods:
...
+// Downloads as 'script' (uses filename exactly as provided)
+editorRef.current.downloadContent('script');
+
+// Downloads as 'my-file.py' (uses filename exactly as provided)
+editorRef.current.downloadContent('my-file.py');
+
+// Downloads as 'code' (uses filename exactly as provided, even though it's "code")
+editorRef.current.downloadContent('code');
+
+// Downloads as 'code.py' (default behavior - no filename parameter provided)
```
Comment (shaneeza) [inline packages/code-editor/README.md]: Is the extension in the comment correct?
